### [`Option`から`Result`を取り出す](#optionからresultを取り出す){.header}

混在するエラー型に対する最も基本的な対処法は、単にお互いを埋め込んでしまうことです。

    use std::num::ParseIntError;

    fn double_first(vec: Vec<&str>) -> Option<Result<i32, ParseIntError>> {
        vec.first().map(|first| {
            first.parse::<i32>().map(|n| 2 * n)
        })
    }

    fn main() {
        let numbers = vec!["42", "93", "18"];
        let empty = vec![];
        let strings = vec!["tofu", "93", "18"];

        println!("The first doubled is {:?}", double_first(numbers));

        println!("The first doubled is {:?}", double_first(empty));
        // Error 1: the input vector is empty
        // エラー１：入力が空ベクトル

        println!("The first doubled is {:?}", double_first(strings));
        // Error 2: the element doesn't parse to a number
        // エラー２：要素が数字としてパースできない
    }

中には、`Option`の中身が`None`の場合はそのまま処理を進め、エラーの検出に限り実行を止めたいという場合もあるでしょう（[`?`](../result/enter_question_mark.html)を使った時のように）。いくつかのコンビネータによって簡単に`Result`と`Option`をスワップすることができます。

    use std::num::ParseIntError;

    fn double_first(vec: Vec<&str>) -> Result<Option<i32>, ParseIntError> {
        let opt = vec.first().map(|first| {
            first.parse::<i32>().map(|n| 2 * n)
        });

        opt.map_or(Ok(None), |r| r.map(Some))
    }

    fn main() {
        let numbers = vec!["42", "93", "18"];
        let empty = vec![];
        let strings = vec!["tofu", "93", "18"];

        println!("The first doubled is {:?}", double_first(numbers));
        println!("The first doubled is {:?}", double_first(empty));
        println!("The first doubled is {:?}", double_first(strings));
    }
:::

